### PR TITLE
User auth 1

### PR DESCRIPTION
**Pull request by jparkSF**
Yo, Check this out!

this is revised ver. of my previous PR

### CODE REVIEW

File: app/controllers/contacts_controller.rb [hunk 1]
```diff
@@ -1,6 +1,11 @@
 class ContactsController < ApplicationController
+
   ACCEPTED_EXTENSIONS = %w(.xlsx .xls .csv).freeze
 
+  before_action :check_admin!, except: [:search]
+  before_action :check_accessibility!, except: [:index, :show, :search]
+  before_action :authenticate_user!
```
Comment (ynghyn) [inline app/controllers/contacts_controller.rb]: I believe `before_action`s are executed in chained order. Let's move this up the chain (ie: to line 4)

File: app/models/user_role.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+class UserRole < ActiveRecord::Base
+
+  belongs_to :user, primary_key: :user_id, foreign_key: :id
+
+  ROLE_TYPE = %w(admin_rwx_access admin_rw_access admin_r_access).freeze
+
+  ADMIN_RWX_ACCESS = 'admin_rwx_access'.freeze
+  ADMIN_RW_ACCESS = 'admin_rw_access'.freeze
```
Comment (ynghyn) [inline app/models/user_role.rb]: should you have a method for this as well?

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable
+
+  has_many :contacts
+  has_one :user_role, primary_key: :id, foreign_key: :user_id
+
+
+  def r_admin?
```
Comment (ynghyn) [inline app/models/user.rb]:
```
def admin?
  user_role.try(:admin_r_access?) || 
    user_role.try(:admin_rw_access?) || 
    user_role.try(:admin_rwx_access?)
end

def rw_admin?
  user_role.try(:admin_rw_access?) || 
    user_role.try(:admin_rwx_access?)
end

def rwx_admin?
  user_role.try(:admin_rwx_access?)
end
```

File: app/models/contact.rb [hunk 4]
```diff
@@ -1,5 +1,6 @@
 class Contact < ActiveRecord::Base
   #before_save :downcase_fields
+  belongs_to :user
```
Comment (ynghyn) [inline app/models/contact.rb]: I don't quite understand the concept of a contact belonging to a user and vice versa. What're we trying to achieve here?

File: app/views/contacts/show.html.erb [hunk 5]
```diff
@@ -75,11 +75,13 @@
     <div id="show-back">
       <%= link_to 'Back', contacts_path, {:class => "button big scrolly",:id => "show_button"} %>
     </div>
+    <% if current_user.id == User.find(1).id %>
```
Comment (ynghyn) [inline app/views/contacts/show.html.erb]:
```
<% if curren_user.rw_admin? %>
  ..show-edit..
<% end %>
<% if curren_user.rwx_admin? %>
  ..show-delete..
<% end %>

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -27,11 +27,22 @@
 
 <body>
   <nav id="nav">
-    <ul class="">
+    <ul>
       <a href="/contacts/"><i id="font-awesome" class="fa fa-home" aria-hidden="true"></i></a>
       <a href="/contacts/search"><i id="font-awesome" class="fa fa-search" aria-hidden="true"></i></a>
-      <a href="/contacts/new"><i id="font-awesome" class="fa fa-plus" aria-hidden="true"></i></a>
-      <a href="/contacts/database"><i id="font-awesome" class="fa fa-database" aria-hidden="true"></i></a>
+      <% if user_signed_in? && current_user.id == 1 %>
```
Comment (ynghyn) [inline app/views/layouts/application.html.erb]: `<% if user_signed_in? && current_user.rw_admin? %>`

File: config/initializers/devise.rb [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+  # confirmation, reset password and unlock tokens in the database.
+  # Devise will use the `secret_key_base` as its `secret_key`
+  # by default. You can change it below and use your own secret key.
+  # config.secret_key = 'e41acc721884916cb86fef614d8be54199f0595633fa05dfaa0e416f29f85bc9833a5fdcddd1d598e8bdf5724a0c6b018873fa9860878f8a0f64fc867fa3a2b0'
+
+  # ==> Mailer Configuration
+  # Configure the e-mail address which will be shown in Devise::Mailer,
+  # note that it will be overwritten if you use your own mailer class
+  # with default "from" parameter.
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
Comment (ynghyn) [inline config/initializers/devise.rb]: update sender

File: config/environments/development.rb [hunk 8]
```diff
@@ -38,4 +38,7 @@
 
   # Raises error for missing translations
   # config.action_view.raise_on_missing_translations = true
+
+
+  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
```
Comment (ynghyn) [inline config/environments/development.rb]: I remember us setting up a mailer server on our Ubuntu box. Anyways, we can double check later and configure this correctly (if not already) so mailing will work.

File: app/controllers/contacts_controller.rb (hunk 1 above)
Comment (jparkSF) [inline app/controllers/contacts_controller.rb]: checked

File: app/views/contacts/show.html.erb (hunk 5 above)
Comment (jparkSF) [inline app/views/contacts/show.html.erb]: good catch. i forgot to rewrite that code. =)

File: app/models/user_role.rb (hunk 2 above)
Comment (ynghyn) [inline app/models/user_role.rb]: dicussed offline. let's just RW

**Pull request merged**